api/src/main/java/com/datatorrent/api/DAG.java

```diff
@@ -250,6 +250,24 @@
   public abstract <T> void setAttribute(Operator operator, Attribute<T> key, T value);
 
   /**
+   * Set affinity between operators
+   * @param locality
+   * @param relaxLocality
+   * @param first operator
+   * @param one or more operators
+   */
+  public abstract void setAffinity(Locality locality, boolean relaxLocality, String firstOperator, String... operators);
```

Comment:
Yes, the affinity/anti-affinity can be set from attributes directly. These were added as convenience methods as per discussion on dev mailing list. Here are sample usages with these methods:
https://github.com/ishark/Apex-Samples/blob/master/affinity-example/src/main/java/com/example/affinity/ApplicationDagAPIs.java#L33-L34

Compared to when setting via attributes:
https://github.com/ishark/Apex-Samples/blob/master/affinity-example/src/main/java/com/example/affinity/ApplicationDagContext.java#L42-L54